Mathlib/Tactic/ExtractGoal.lean

```diff
@@ -116,7 +117,11 @@ elab (name := extractGoal) "extract_goal" full?:&"*"? name?:(colGt ppSpace ident
              else mkAuxName ((← getCurrNamespace) ++ `extracted) 1
   let msg ← withoutModifyingEnv <| withoutModifyingState do
     let mut g ← getMainGoal
-    unless full?.isSome do
+    let mut cleanup := full?.isNone
+    if (← g.getType >>= instantiateMVars).consumeMData.isConstOf ``False then
+      -- In a contradiction proof, it is not very helpful to clear all hypotheses!
+      cleanup := false
+    if cleanup then
```

Comment:
Isn't this a convoluted way of doing
```
if full?.isNone && !((← g.getType >>= instantiateMVars).consumeMData.isConstOf ``False) then
      g ← g.cleanup

Reply:
Or even better:
```
unless full?.isSome || (← g.getType >>= instantiateMVars).consumeMData.isConstOf ``False do
      g ← g.cleanup
```
although it would be nice to make sure this short-circuits when `full?.isSome` and there is a trap here if I remember correctly.

Reply:
I think to avoid the trap you need `... <||> do ...`, and forgetting the `do` springs the trap.

Reply:
Maybe it'll make you question my tastes, but, believe it or not, I considered all three of these options and settled on using a `mut` variable. I'd thought that maybe there might be other forms we'd want to match on eventually and this structure is easier to read than an `||` chain.

Anyway, now the convolution is gone so we don't need to worry about it anymore :smile:

Reply:
I don't mind the convoluted way, I mostly wanted to check I wasn't missing a learning opportunity.

Reply:
I probably should have just done the `if ... &&` or the `unless ... ||` from the beginning -- the likelihood of needing to match anything else is really low. (Btw, I'd decided that the "`do` trap" was ok here because instantiating mvars and doing `isConstOf` is inexpensive. If it were more expensive, I'd do `... <||> do ...`.)